### ISSUE TITLE
Demos / Browser-Based Usage

### ISSUE THREAD

**Issue by githubpsyche** [HLasse/TextDescriptives#212]
Hi, thanks for developing this.

IMO the sustained popularity of text descriptive tools like Coh-Metrix is/was driven in part by the way they were usable from the browser without any programming. While Coh-Metrix was supported, people could go to a website, paste their text into a textbox, and after a small wait would download a table containing all the metrics they needed. It might serve the popularity of this tool if there were a similar website for generating TextDescriptives outputs.

Relatedly, I made a colab notebook for a collaborator to serve a similar purpose, linked at the bottom of this post. Maybe a cleaned-up version of this is a lofi way to achieve the same result. Or maybe something hosted on a Huggingface space is a better home for an interactive demo; idk. https://huggingface.co/spaces/launch

https://colab.research.google.com/drive/1hhf2skio2q7nnD8lMRXCfYBPvA4mG-PE?usp=sharing

**Comment by KennethEnevoldsen** [comment]
Thanks @githubpsyche, I totally agree that would def. be a very user-friendly addition to TextDescriptives.

@HLasse or @LudvigOlsen, is this something any of you have time for? or @githubpsyche if you wish to contribute a pull request would also be a very welcome

**Comment by HLasse** [comment]
Great suggestion, I can definitely see the usefulness for a broader digital humanities crowd. I don't have the time to look into this before July/August, so a PR would be very welcome. @LudvigOlsen, do you have any experience with this type of development? 

Seems like this could be done fairly easily as a Gradio/streamlit app which would also mean that we don't have to host anything ourselves.

**Comment by LudvigOlsen** [comment]
Developing this with Streamlit would probably be quite easy yes. It is very easy to make something quick in. If hugging face can host it even, then it shouldn't be a problem to do.

**Comment by LudvigOlsen** [comment]
I'll try to make something simple while watching chess :-) No promises

**Comment by HLasse** [comment]
No experience with Gradio either, but seems like Streamlit is further along in development. Completely up to you, thanks a lot for looking at this Ludvig! No pressure, looking forward to seeing it

**Comment by LudvigOlsen** [comment]
Here's a first try for a streamlit application that allows inputting text by 1) text box, 2) upload of a .txt file. 

https://github.com/LudvigOlsen/textdescriptives_app

Please try it out locally and give some feedback.

Perhaps it's best to transfer the repo to you @HLasse so you can keep the two together? 

If @HLasse or @KennethEnevoldsen has time to look at the huggingface "hosting", please go ahead. I likely won't have a lot of time in the coming days.

**Comment by HLasse** [comment]
Really nice work Ludvig! Since it's streamlit, we can use streamlit's own hosting service which [seems fairly straightforward](https://docs.streamlit.io/streamlit-community-cloud/get-started/deploy-an-app). I think huggingface is mainly geared towards Gradio apps, but I haven't checked tbh. 

Perhaps we should add:
- Dropdown for language
- Dropdown for model size
- Some way to store the models on the hosting server so they don't have to be redownloaded every time

But excellent start! 

@githubpsyche, are there particular features from e.g. Coh-Metrix that you'd like to see implemented? I'm not familiar with it. 

Here's a screenshot of the current app
<img width="757" alt="image" src="https://user-images.githubusercontent.com/23191638/232799497-848583f2-d15f-4416-b4e7-9d375c8495ef.png">

**Comment by LudvigOlsen** [comment]
I've added the submit ("apply") button so the user can change multiple settings before running the extraction!

**Comment by LudvigOlsen** [comment]
New screenshot (13" screen lol)

<img width="757" alt="billede" src="https://user-images.githubusercontent.com/22819047/232805577-d6108fc0-b5e3-4298-a082-6befee9ccda1.png">

**Comment by githubpsyche** [comment]
I think this app looks great. Definitely matches my memory of coh metrix's functionality. I'm looking forward to sharing it with people!

Jordan
________________________________

**Comment by KennethEnevoldsen** [comment]
@LudvigOlsen I can solve the hosting. I recall that it was quite easy to host it on streamlit. Unless we have a strong preference for Huggingface spaces? (probably more exposure there)

**Comment by KennethEnevoldsen** [comment]
So reading up on the streamlit hosting, we can simply add it to this repository here in a `.streamlit`-folder. That is probably. Then we can move the discussion of the interface over to a PR instead

Looks great btw!

**Comment by githubpsyche** [comment]
I suspect a HF space would be good for exposure, but I don’t really know - people browsing that site are usually looking for ML models. Could always do both? A HF space is free if I remember correctly.

Jordan
________________________________

**Comment by KennethEnevoldsen** [comment]
Hmm a huggingface space is github repo with an app.py script. That would require us to have two places where we maintain and app (which might have different type of errors). I would prefer only hosting it in one space.

**Comment by LudvigOlsen** [comment]
@KennethEnevoldsen I saw an old blog post that mentioned a 1gb storage cap for streamlit apps (free accounts). Don't know whether that still is the case but perhaps we should 1) find out what the limitations are, 2) calculate if we can stay within those limitations. I'm not sure how big the models are, but if you have multiple users at a time uploading large text files and loading the models into ram, perhaps huggingface is better suited for that kind of application?
If these limitations are not problematic, having the app in this repository does seem easier yes!

@HLasse 
Perhaps we should add some kind of reproducibility feature where the current package version tag is mentioned in the app for better citations? And perhaps add the option to download a script that runs the exact same code calling the package?

**Comment by LudvigOlsen** [comment]
I've added the package version to the introduction text (based on `td.__version__`):

<img width="715" alt="billede" src="https://user-images.githubusercontent.com/22819047/232963813-0037d217-d1e3-4c05-a007-40a57bf26dcb.png">

I've also added an expandable section with the python code. This serves both as educational content and for reproducibility. Of course it needs to be maintained when we add options, but at this point the application is so simple that it shouldn't be a problem:

<img width="730" alt="billede" src="https://user-images.githubusercontent.com/22819047/232964007-46358ef0-83d7-4212-beb2-ccebc02bda55.png">

I further cleaned up the code a bit.

**Comment by HLasse** [comment]
Excellent work, Ludvig!! I think all that's left is to add a language and a model size toggle instead of the model name and then figuring out the best way to host. 

1 GB ram is probably going to be quite limiting - IIRC spacy large models are like 500 MB so if there's more than one concurrent user we'll quickly run into problems. Huggingface space has vastly more (see screenshot), but I completely get the point that maintaining at multiple places is annoying. 
<img width="469" alt="image" src="https://user-images.githubusercontent.com/23191638/233024722-ce13404d-9871-42a2-9912-4b55b61d0ce7.png">

Alternatively, this might be something that CHC might want to host/provide as a service? What do you think, @KennethEnevoldsen? I can reach out to Kristoffer/Peter to hear their thoughts as well

**Comment by LudvigOlsen** [comment]
Will you add those choices? Otherwise, let me know the options and I'll add the dropdowns.

Probably nice to have a platform host. Reduces complexity for us I assume? I'm indifferent though, so you guys just figure out what you prefer!

**Comment by HLasse** [comment]
You're most welcome to add them! :D 

- A dropdown for the language (one of these: https://spacy.io/usage/models)
- a dropdown for model size (options: sm, md, lg, trf)
- a dropdown for metrics (with multi select). Probably default to all of them.

**Comment by LudvigOlsen** [comment]
I don't see the point in not calculating all metrics in this interface? It's easy to remove columns afterwards and selecting from a dropdown seems annoying when are this many options.

**Comment by LudvigOlsen** [comment]
Could you specify the languages you would like to include? Some seem to not have models and some seem to have extra dependencies? If you could write me a dict mapping "Full language name" -> "language code", it will be easy for me to add.

**Comment by LudvigOlsen** [comment]
I tried selecting the transformer model for Danish. It first installs a lot of stuff (we should have these pre-installed in the environment, if we wish to support this model type) and then fails with:

```
Can't find factory for 'transformer' for language Danish (da). 
This usually happens when spaCy calls `nlp.create_pipe` with a custom component name that's not registered on the current language class. If you're using a Transformer, make sure to install 'spacy-transformers'. 
If you're using a custom component, make sure you've added the decorator `@Language.component` (for function components) or `@Language.factory` (for class components).
```

So if some combinations of language and model size are not allowed, we should probably handle that somehow?

**Comment by LudvigOlsen** [comment]
Actually I get a similar error for the basic case (English and small), when selecting the `pos_stats` but not when supplying `None` (which yields pos metrics):

<img width="581" alt="billede" src="https://user-images.githubusercontent.com/22819047/233040666-c8211ba9-32e4-46f3-a28a-f704b011bce4.png">

Do we really need to support selection of metrics? It does explain what categories are available but also clutters the interface:

<img width="741" alt="billede" src="https://user-images.githubusercontent.com/22819047/233041999-c62b4474-5a71-4288-b3cd-6e81a371df2b.png">

**Comment by LudvigOlsen** [comment]
I've disabled metrics selection for now. You can enable it by changing a hardcoded `False` to `True` in main.py. There's a new file called `options.py`, where you can add the options to the dicts, and it should work. The code has been pushed.

**Comment by LudvigOlsen** [comment]
Perhaps we should add a link to explanations of the metrics?

**Comment by HLasse** [comment]
Awesome work Ludvig! I think it's useful to be able to toggle metrics as some of them take a while to calculate (e.g. quality) and might be fairly niche. So if you're only interested in say readability you can save a lot of time by disabling unwanted metrics. 

The second error is because the pos_stats component is actually called `pos_proportions`. 

The first one is a bit more tricky. There's a json file here which lists the available models for each spacy version: https://github.com/explosion/spacy-models/blob/master/compatibility.json 

Perhaps we can parse this and check whether the model is valid. This should probably be done in [_download_spacy_model](https://github.com/HLasse/TextDescriptives/blob/a78c863452a18f44e749d7409cf2460f4cb7be67/src/textdescriptives/utils.py#L85) though. Example to start from [here](https://github.com/explosion/spaCy/issues/4592#issuecomment-549900338)

Not sure why the transformer model fails for Danish though, that seems odd. I'll take a look at it.

> Perhaps we should add a link to explanations of the metrics?

+1 . Perhaps just a link to the [docs](https://hlasse.github.io/TextDescriptives/index.html)? Everything is explained under the "Components" header

Re the language dict, I wonder if spaCy doesn't keep one somewhere. Couldn't find it browsing through their repo though. @KennethEnevoldsen , have you stumbled upon a list of supported languages anywhere?

**Comment by LudvigOlsen** [comment]
Fair. I was also looking at the function documentation for the option name, which is obviously wrong of me :D 
Made a PR (https://github.com/HLasse/TextDescriptives/pull/214) to fix it :-)

I guess it should be possible to parse the json file yes. Seems annoying to do on the fly though - although that would keep it up to date automatically. Perhaps it could be reparsed once a week or something with GH actions?

> +1 . Perhaps just a link to the [docs](https://hlasse.github.io/TextDescriptives/index.html)? Everything is explained under the "Components" header

Yeah. Probably the easiest approach.

A spacy language dict would be the optimal solution.

**Comment by LudvigOlsen** [comment]
Here's an approach to checking availability.

```python
from spacy.cli.download import get_compatibility

def available_models() -> List[str]:
    return list(get_compatibility().keys())

def is_available(lang: str, size: str) -> bool:
    lang_and_size = set([
        "_".join(list(map(m.split("_").__getitem__, [0,-1])))
        for m in available_models()
    ])
    return f"{lang}_{size}" in lang_and_size

```

I think `get_compatibility()` simply gets the part of the json file relevant the the installed `spacy` version. 

Will use a version of this to filter the model size options available to the language.

**Comment by HLasse** [comment]
That looks great! Perhaps we shouldn't give users the option of the transformer (trf) models - we're not going to be hosting a GPU instance so it'll probably be pretty slow (and expensive) compared to the other options

**Comment by LudvigOlsen** [comment]
I have removed the transformer version for now.

I can't update the model size options based on the selected language as there's no rerendering inside forms before pressing "apply". But I've made this check when submitting that should be better for now:

<img width="743" alt="billede" src="https://user-images.githubusercontent.com/22819047/233096240-a820c0f1-9a1e-4380-8fa5-e9605f9030d1.png">

**Comment by HLasse** [comment]
That's fantastic!! What's left then? A mapping from language to language code?

**Comment by HLasse** [comment]
Alrighty, we got a demo up and running! https://huggingface.co/spaces/HLasse/textdescriptives?fbclid=IwAR3CA5li61K0YQEuEzt4EsoiFB7QSQ7JFfjGjl3RqMVat9G_SQd6RUR1Ou0

Spread the word and provide feedback, please! I'll add a badge in the readme as well.

Thanks a ton for coding this up, @LudvigOlsen!

**Comment by KennethEnevoldsen** [comment]
I think this looks great! I have added a PR to add it a few more places to the documentation. @githubpsyche please let us know if there is any feedback on the functionality of the app. 

Once the PR is merged I think we can close this PR. Great work!
